### ISSUE TITLE
GPU backend (WebGPU/OpenCL via ml_drift) silently omits trailing generated content after a tag-like boundary; CPU backend produces correct output on identical model

### ISSUE THREAD

**Issue by bosejapan23-06** [google-ai-edge/litert-torch#1148]
## Summary
On the GPU backend, a fine-tuned Qwen2.5-1.5B model trained to produce
structured output in the form `<MOOD>state</MOOD> <TOOL>toolName(args)</TOOL>`
frequently stops generation immediately after the first tag
(`</MOOD>`), omitting the second tag (`<TOOL>...</TOOL>`) that the
identical .litertlm file reliably produces on the CPU backend for the same
input. Across a 60-case evaluation suite, GPU tool-tag correctness is
35/60 vs. 56/60 on CPU -- same model file, same prompts.

## Environment
- litert-lm: 0.14.0
- litert-torch-nightly: (please substitute your currently installed version
  via `pip show litert-torch-nightly`)
- Backend tested: GPU (WebGPU/OpenCL via ml_drift), compared against CPU
- Model: fine-tuned Qwen2.5-1.5B-Instruct, LoRA-merged, exported via
  litert-torch export_hf with the StableHLOCompositeBuilder neutralization
  patch (see issue #1136) applied, quantization_recipe=dynamic_wi8_afp32,
  prefill_lengths=[32,128,512,1280]

## Steps to Reproduce
1. Export a fine-tuned Qwen2.5-1.5B model with the above configuration.
2. Run identical prompts through both backends:
   litert-lm run model.litertlm --backend=cpu --prompt="<real prompt>"
   litert-lm run model.litertlm --backend=gpu --prompt="<real prompt>"
3. Compare outputs across many cases where the expected output includes
   two sequential tags.

Example (identical prompt, both backends):
  CPU: "Of course, I'm warming it up for you. <MOOD>happy</MOOD> <TOOL>increaseTemperature(all)</TOOL>"
  GPU: "Of course, I'm warming it up for you. <MOOD>happy</MOOD>"

## What we've ruled out
- Random sampling: confirmed deterministic via repeated identical runs.
- Generation-length limit: --max-num-tokens is documented as the KV-cache
  size, not an output-length limiter; confirmed no change when explicitly
  maximized.
- Speculative decoding: --enable-speculative-decoding=false produces
  byte-identical output to the default; ruled out.
- Quantization recipe: tested weight_only_wi8_afp32 (different, unrelated
  crash), dynamic_wi8b32_afp32 and dynamic_wi8b64_afp32 (both silently
  apply zero quantization in our installed version -- Quantization Ratio:
  1.00 in the export log -- so not a valid data point), and
  dynamic_wi8c_afp32 (per-channel granularity -- applies real, correct
  quantization at the expected size, but exhibits the identical bug).
  The one configuration that DID produce correct GPU output used no
  quantization at all (full fp32/bf16, ~6.6GB, not deployable at that
  size), which is consistent with -- but not proof of -- a
  quantization-precision-related root cause specifically in GPU kernel
  execution.
- Prefill bucket structure: tested a tighter-fitting bucket set: no change.

## Expected
Identical (or functionally equivalent) output between CPU and GPU backends
for the same deterministic generation request.

## Actual
GPU frequently terminates generation right after the first tag closes,
even though the model's own reasoning (the preceding sentence) clearly
indicates it intended to continue. CPU produces the complete, correct
output for the identical input every time in our test suite.

## Hypothesis
Given content and mood are consistently correct and only this one specific
decode-position decision differs, we suspect the GPU delegate's floating-
point computation (different accumulation order/kernel implementation vs.
CPU) produces a logit distribution that's sufficiently different from CPU's
at this one close-call decision point (continue vs. end-of-sequence) to
flip the argmax outcome, even though both backends agree confidently
everywhere else in the sequence.

## Question
Is this a known characteristic of the ml_drift GPU delegate's numerical
precision vs. the CPU/XNNPack path? Is there a recommended way to bias
generation to reduce sensitivity to this kind of near-tie at tag/structural
boundaries, or a delegate-level fix planned?

Happy to provide the exact model export command, full evaluation logs, or
run additional targeted reproduction steps if useful.

**Comment by shivvaam0001** [comment]
Hi @bosejapan23-06,

Thank you for reaching out to us.
Could you please provide the model export command, and relevant logs so that we can reproduce and investigate the issue on our end?

Thanks!

**Comment by bosejapan23-06** [comment]
hi @shivvaam0001 ,
Thanks for picking this up! Here's the exact export command

`from types import SimpleNamespace
from litert_torch.generative.export_hf.core import attention
class _PassthroughBuilder:
    def __init__(self, *a, **k): pass
    def mark_inputs(self, *xs): return xs[0] if len(xs)==1 else xs
    def mark_outputs(self, *xs): return xs[0] if len(xs)==1 else xs
attention.composite = SimpleNamespace(StableHLOCompositeBuilder=_PassthroughBuilder)
from litert_torch.generative.export_hf import export as export_module
export_module.export(
    model='outputs/qwen_shrunk_v2_merged',  # LoRA-merged Qwen2.5-1.5B-Instruct
    output_dir='outputs/litert_qwen_shrunk_v2_gpu',
    externalize_embedder=True,
    prefill_lengths=[32,128,512,1280],
    cache_length=4096,
    use_jinja_template=True,
    quantization_recipe='dynamic_wi8_afp32',
    enable_gpu_dynamic_prefill=True,
    enable_gpu_dynamic_cache=True,
    llm_metadata_max_num_tokens_override=4096,
)`

(The StableHLOCompositeBuilder monkeypatch neutralizes the odml.softmax composite marker from issue #1136 -- required for this model to delegate to GPU at all. model is a standard HF-format LoRA-merged Qwen2.5-1.5B-Instruct checkpoint, no unusual config.)

**Offer to share the actual model file**

Since this is a fine-tuned model, an exact reproduction needs the same weights -- happy to share the merged .litertlm (~1.8GB) or the pre-export HF-format merged checkpoint via a private link if that's useful, rather than you having to approximate the fine-tune. Let me know which format is more convenient and where to send it.

**Representative log -- full EngineSettings + side-by-side output**

This is from the Python litert-lm-api package, comparing default GPU settings against explicitly forcing activation_data_type=ActivationDataType.FLOAT32 (same model, same prompt, same sampler config, only that one field changed).

Expected: forcing FP32 activation type produces different (correct) output than the FP16 default, or at minimum the log confirms FP32 is actually being used for decode.

Actual: the runtime's own log confirms the default is FP16 for the text decoder specifically -- and forcing FP32 via this parameter does not change the generated output at all.

<details> <summary>Full log excerpt (click to expand)</summary>
--- Run 1: activation_data_type=ActivationDataType.FLOAT32 ---

W0000 00:00:1785731688.473981    6737 litert_lm_loader.h:174] TFLite model
type: TF_LITE_VISION_ENCODER not found for prefer activation type. Use
system's default backend activation type. System's default activation type
for Text decoder is fp16. Vision encoder and audio encoder default is fp32.

I0000 00:00:1785731688.475008    6737 engine_settings.cc:448] The validated
engine settings: EngineSettings:
  MainExecutorSettings: backend: GPU
backend_config:
max_top_k: 1
max_tokens: 4096
activation_data_type: FLOAT32
...
enable_speculative_decoding: false
...

=== FP32 (test) ===
Of course, I'm warming it up for you. <MOOD>happy</MOOD>


--- Run 2: activation_data_type not set (default) ---

W0000 00:00:1785731694.723497    6737 litert_lm_loader.h:174] TFLite model
type: TF_LITE_VISION_ENCODER not found for prefer activation type. Use
system's default backend activation type. System's default activation type
for Text decoder is fp16. Vision encoder and audio encoder default is fp32.

I0000 00:00:1785731694.723512    6737 engine_settings.cc:448] The validated
engine settings: EngineSettings:
  MainExecutorSettings: backend: GPU
backend_config:
max_top_k: 1
max_tokens: 4096
activation_data_type: Not set
...
enable_speculative_decoding: false
...

=== default ===
Of course, I'm warming it up for you. <MOOD>happy</MOOD>
</details>

Note the activation_data_type field is confirmed different between the two runs (FLOAT32 vs Not set) -- so the parameter is being received and threaded through correctly -- but the generated text is byte-identical either way. Both runs also show the runtime explicitly stating: "System's default activation type for Text decoder is fp16." -- confirming FP16 is the actual default for GPU text decoding, independent of whatever this particular API parameter controls.

**Comment by aravindmurali22** [comment]
Hi @protobird-git , following up on this thread, is there any update on this issue? Thanks!
